### PR TITLE
De-structure into smaller files and add more assertions

### PR DESCRIPTION
**Pull request by mroderick**
This PR is in preparation of tidying the codebase up, so we can add a plugin system, to make it much more convenient to make custom assertions for referee.

#### New assertions

I am adding a bunch of new assertions, basically wanting to cover all the built in types in JavaScript (only final versions).

#### How to verify
1. Check out this branch
2. `npm install`
3. `npm test`

#### Checklist for author

- [x] `npm run lint` passes
- [x] References to standard library functions are [cached](https://github.com/sinonjs/sinon/pull/1523).

### CODE REVIEW

**Comment by mroderick** [comment]
Ping @grommett

File: lib/assertions/is-array-like.js [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+
+module.exports = function (referee) {
+    function isArrayLike(object) {
+        return Array.isArray(object) ||
+            (!!object && typeof object.length === "number" &&
+            typeof object.splice === "function") ||
+            isArguments(object);
+    }
+
+    referee.isArrayLike = isArrayLike;
```
Comment (grommett) [inline lib/assertions/is-array-like.js]: Is this necessary? This seems unlike the previous plugins. In this case you're writing directly to `referee` as opposed to using the `referee.add` interface. Tests pass when removing this line.

File: lib/referee.js [hunk 2]
```diff
@@ -211,499 +181,41 @@ referee.prepareMessage = function msg(message) {
...
+    "./assertions/is-object",
+    "./assertions/is-string",
+    "./assertions/is-true",
+    "./assertions/keys",
+    "./assertions/less",
+    "./assertions/match",
+    "./assertions/near",
+    "./assertions/same",
+    "./assertions/tag-name"
+].forEach(function (config) {
```
Comment (grommett) [inline lib/referee.js]: Nice!!

File: lib/assertions/is-array-like.js (hunk 1 above)
Comment (mroderick) [reply lib/assertions/is-array-like.js]: I am trying not to break the public API ... yet, so we can still merge in all these changes to `master`, and still make a release.

**Comment by grommett** [comment]
👍

**Comment by mroderick** [comment]
@mantoni @grommett I am done added changes to this PR, your reviews would be most welcome

**Comment by grommett** [comment]
👍

**Comment by mroderick** [comment]
Thanks

**Comment by mroderick** [comment]
This has been published as `@sinonjs/referee@2.1.0`

**Pull request merged**